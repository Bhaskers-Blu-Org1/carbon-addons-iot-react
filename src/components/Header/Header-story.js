import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { Icon } from 'carbon-components-react';
import { rem } from 'polished';
import styled from 'styled-components';

import Header from './Header';

React.Fragment = ({ children }) => children;

const User = styled.p`
   {
    color: white;
    font-size: 0.75rem;
    text-align: left;
    margin-right: ${rem(20)};
  }
`;

const DropItem = styled.span`
   {
    display: flex;
  }
`;

const StyledIcon = styled(Icon)`
   {
    width: 25px;
    height: 25px;
  }
`;

// const link = <Icon name="header--help" fill="white" description="Icon" />;
const HeaderProps = {
  user: 'JohnDoe@ibm.com',
  tenant: 'TenantId: Acme',
  className: 'custom-class-name',
  appName: 'Watson IoT Platform ',
  actionItems: [
    {
      label: 'alerts',
      onClick: action('pop alert modal'),
      btnContent: <StyledIcon name="notification-on" fill="white" description="Icon" />,
    },
    {
      label: 'help',
      onClick: action('help drop open'),
      btnContent: (
        <StyledIcon
          name="header--help"
          fill="white"
          description="Icon"
          className="bx--header__menu-item bx--header__menu-title"
        />
      ),
      childContent: [
        {
          onCLick: action('go to source'),
          content: <p>This is a link</p>,
        },
        {
          onCLick: action('go to source'),
          content: (
            <React.Fragment>
              <DropItem>
                JohnDoe@ibm.com
                <StyledIcon name="header--avatar" fill="white" description="Icon" />
              </DropItem>
            </React.Fragment>
          ),
        },
      ],
    },
    {
      label: 'user',
      onClick: action('click'),
      btnContent: (
        <React.Fragment>
          <User>
            JohnDoe@ibm.com<span>TenantId: Acme</span>
          </User>
          <StyledIcon name="header--avatar" fill="white" description="Icon" />
        </React.Fragment>
      ),
    },
  ],
};

storiesOf('Header', module).add('Header action buttons with dropdowns', () => (
  <Header {...HeaderProps} />
));
