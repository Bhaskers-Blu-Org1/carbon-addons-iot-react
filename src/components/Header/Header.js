import { Icon } from 'carbon-components-react';
import {
  Header as CarbonHeader,
  HeaderName,
  HeaderGlobalBar,
  HeaderGlobalAction,
  // HeaderMenuItem,
  // HeaderNavigation,
  // HeaderMenuButton,
} from 'carbon-components-react//lib/components/UIShell';
import { rem } from 'polished';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import React from 'react';

import { COLORS } from '../../styles/styles';

const StyledHeader = styled(CarbonHeader)`
   {
    background-color: ${COLORS.darkGray};
    margin-top: 2rem;
  }
`;
const StyledGlobalAction = styled(HeaderGlobalAction)`
   {
    align-items: center;
    display: flex;
    justify-content: center;
    min-width: 3rem;
    padding: 0 ${rem(15)};
    width: auto;
    :hover {
      background-color: ${COLORS.darkGrayHover};
    }
    :last-child svg {
      margin-left: ${rem(20)};
    }
    span {
      display: block;
    }
  }
`;
const User = styled.p`
   {
    color: white;
    font-size: 0.75rem;
    text-align: left;
  }
`;
const StyledIcon = styled(Icon)`
   {
    width: 25px;
    height: 25px;
  }
`;

const propTypes = {
  onClick: PropTypes.func.isRequired,
  /** Shows current user */
  user: PropTypes.string.isRequired,
  /** Shows tenant or Org ID */
  tenant: PropTypes.string.isRequired,
  /** Name ot follow the IBM prefix up top, left */
  appName: PropTypes.string.isRequired,
};

/**
 * Clickable card that shows "Add" button
 */
const Header = ({ appName, className, user, tenant, onClick }) => (
  <StyledHeader className={className} aria-label="main header" onClick={onClick}>
    {/* <HeaderMenuButton aria-label="Open menu" onClick={() => true} /> */}
    <HeaderName href="#" prefix="IBM">
      {appName}
    </HeaderName>
    {/* <HeaderNavigation aria-label="IBM [Platform]">
      <HeaderMenuItem element={linkElement}>Catalog</HeaderMenuItem>
    </HeaderNavigation> */}
    <HeaderGlobalBar>
      <StyledGlobalAction aria-label="Profile" onClick={() => true}>
        <StyledIcon name="header--help" fill="white" description="Icon" />
      </StyledGlobalAction>
      <StyledGlobalAction aria-label="Profile" onClick={() => true}>
        <User>
          <span>{user}</span>
          {tenant}
        </User>
        <StyledIcon name="header--avatar" fill="white" description="Icon" />
      </StyledGlobalAction>
    </HeaderGlobalBar>
  </StyledHeader>
);

Header.propTypes = propTypes;

export default Header;
