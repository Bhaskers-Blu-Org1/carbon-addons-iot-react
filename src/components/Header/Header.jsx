import {
  Header as CarbonHeader,
  HeaderMenuButton,
  HeaderName,
  HeaderGlobalBar,
  HeaderGlobalAction,
  SkipToContent,
  HeaderMenuItem,
  HeaderMenu,
} from 'carbon-components-react/lib/components/UIShell';
import { rem } from 'polished';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import React from 'react';

import { COLORS } from '../../styles/styles';

// import HeaderMenu from './HeaderMenu';

const StyledHeader = styled(CarbonHeader)`
  &&& {
    .bx--skip-to-content:focus {
      min-width: 240px;
      justify-content: center;
      border: 0.125rem solid ${COLORS.white};
    }

    .bx--header__menu-toggle {
      display: block;
    }

    .bx--header__menu-title[role='menuitem'][aria-expanded='true'] + .bx--header__menu {
      left: auto;
      right: 0;
    }

    .bx--header__menu {
      min-width: 12.5rem;
      width: auto;
    }

    .bx--header__menu-item[role='menuitem']:focus {
      border-color: ${COLORS.white};
      outline: none;
    }
  }
`;
const StyledGlobalAction = styled(HeaderGlobalAction)`
  &&& {
    align-items: center;
    display: flex;
    justify-content: center;
    min-width: 3rem;
    padding: 0 ${rem(15)};
    position: relative;
    width: auto;

    span {
      display: flex;
    }
  }
`;

const propTypes = {
  /** Add a prefix other than IBM */
  prefix: PropTypes.string,
  /** Name to follow the IBM prefix up top, left */
  appName: PropTypes.string.isRequired,
  /** Add a class name to Header */
  className: PropTypes.string,
  /** Provide ID for the skip to content functionality */
  skipto: PropTypes.string,
  /** Object of action items */
  actionItems: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      onClick: PropTypes.func,
      btnContent: PropTypes.any.isRequired,
      childContent: PropTypes.arrayOf(
        PropTypes.shape({
          /** extra data to pass to HeaderMenuLink (aria-*, href, target, etc.) */
          metaData: PropTypes.object,
          /** Optionally pass in an onClick handler to trigger action  */
          onClick: PropTypes.func,
          content: PropTypes.any.isRequired,
        })
      ),
    })
  ).isRequired,
  /** Bit to flip that tells header to render the nav toggle button */
  hasSideNav: PropTypes.bool,
  onClickSideNavExpand: PropTypes.func,
};

const defaultProps = {
  onClickSideNavExpand: null,
  hasSideNav: true,
  prefix: 'IBM',
  className: 'main-header',
  skipto: '#main-content',
};

/**
 * Clickable card that shows "Add" button
 */
const Header = ({
  appName,
  className,
  actionItems,
  prefix,
  skipto,
  hasSideNav,
  onClickSideNavExpand,
}) => {
  const actionBtnContent = actionItems.map(item => {
    if (item.hasOwnProperty('childContent')) {
      const children = item.childContent.map(childItem => (
        <HeaderMenuItem
          key={`menu-item-${item.label + item.childContent.indexOf(childItem)}-child`}
          {...childItem.metaData}
        >
          {childItem.content}
        </HeaderMenuItem>
      ));
      return (
        <HeaderMenu
          key={`menu-item-${item.label}`}
          aria-label={item.label}
          isMenu={false}
          renderMenuContent={() => item.btnContent}
          menuLinkName={item.label}
        >
          {children}
        </HeaderMenu>
      );
    }
    return (
      <StyledGlobalAction
        key={`menu-item-${item.label}-global`}
        aria-label={item.label}
        onClick={item.onClick}
      >
        {item.btnContent}
      </StyledGlobalAction>
    );
  });

  return (
    <StyledHeader className={className} aria-label="main header">
      <SkipToContent href={skipto} />
      {hasSideNav && <HeaderMenuButton aria-label="Open menu" onClick={onClickSideNavExpand} />}
      <HeaderName href="#" prefix={prefix}>
        {appName}
      </HeaderName>
      <HeaderGlobalBar>{actionBtnContent}</HeaderGlobalBar>
    </StyledHeader>
  );
};

Header.propTypes = propTypes;
Header.defaultProps = defaultProps;

export default Header;
