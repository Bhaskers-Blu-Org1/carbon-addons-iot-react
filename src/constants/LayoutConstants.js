export const COLORS = {
  RED: 'rgb(217,3,103)',
  BLUE: 'rgb(25,98,253)',
  YELLOW: 'rgb(254,213,34)',
  PURPLE: '#8A3FFC',
  TEAL: '#009C98',
  MAGENTA: '#EE538B',
  CYAN: '#0072C3',
};

export const CARD_SIZES = {
  XSMALL: 'XSMALL',
  XSMALL_WIDE: 'XSMALL_WIDE',
  SMALL: 'SMALL',
  TALL: 'TALL',
  MEDIUM: 'MEDIUM',
  WIDE: 'WIDE',
  LARGE: 'LARGE',
  XLARGE: 'XLARGE',
};

export const CARD_TYPES = {
  TIMESERIES: 'TIMESERIES',
  VALUE: 'VALUE',
  TABLE: 'TABLE',
  DONUT: 'DONUT',
  PIE: 'PIE',
  BAR: 'BAR',
};

export const DASHBOARD_SIZES = {
  XLARGE: 'xl',
  LARGE: 'lg',
  MEDIUM: 'md',
  SMALL: 'sm',
  XSMALL: 'xs',
};

export const DASHBOARD_COLUMNS = {
  xl: 12,
  lg: 12,
  md: 12,
  sm: 6,
  xs: 4,
};

export const DASHBOARD_BREAKPOINTS = {
  xl: 1800,
  lg: 1056,
  md: 800,
  sm: 480,
  xs: 320,
};

export const ROW_HEIGHT = {
  max: 128,
  xl: 128,
  lg: 128,
  md: 128,
  sm: 128,
  xs: 128,
};

/** The amount of space to preserve between cards */
export const GUTTER = 16;

/*
 * How many rows and columns should each card cross at each layout size
 */

export const CARD_DIMENSIONS = {
  XSMALL: {
    xl: { w: 1, h: 1 },
    lg: { w: 2, h: 1 },
    md: { w: 3, h: 1 },
    sm: { w: 2, h: 1 },
    xs: { w: 2, h: 1 },
  },
  XSMALL_WIDE: {
    xl: { w: 3, h: 1 },
    lg: { w: 4, h: 1 },
    md: { w: 4, h: 1 },
    sm: { w: 6, h: 1 },
    xs: { w: 4, h: 1 },
  },
  SMALL: {
    xl: { w: 2, h: 2 },
    lg: { w: 3, h: 2 },
    md: { w: 4, h: 2 },
    sm: { w: 3, h: 2 },
    xs: { w: 4, h: 2 },
  },
  TALL: {
    xl: { w: 2, h: 4 },
    lg: { w: 3, h: 4 },
    md: { w: 4, h: 4 },
    sm: { w: 3, h: 4 },
    xs: { w: 4, h: 4 },
  },
  MEDIUM: {
    xl: { w: 4, h: 2 },
    lg: { w: 6, h: 2 },
    md: { w: 8, h: 2 },
    sm: { w: 6, h: 2 },
    xs: { w: 4, h: 2 },
  },
  WIDE: {
    xl: { w: 6, h: 2 },
    lg: { w: 9, h: 2 },
    md: { w: 12, h: 2 },
    sm: { w: 6, h: 2 },
    xs: { w: 4, h: 2 },
  },
  LARGE: {
    xl: { w: 4, h: 4 },
    lg: { w: 6, h: 4 },
    md: { w: 6, h: 4 },
    sm: { w: 6, h: 4 },
    xs: { w: 4, h: 4 },
  },
  XLARGE: {
    xl: { w: 6, h: 4 },
    lg: { w: 12, h: 4 },
    md: { w: 12, h: 4 },
    sm: { w: 6, h: 4 },
    xs: { w: 4, h: 4 },
  },
};

export const CARD_LAYOUTS = {
  HORIZONTAL: 'HORIZONTAL',
  VERTICAL: 'VERTICAL',
};

export const CARD_TITLE_HEIGHT = 48;
export const CARD_CONTENT_PADDING = 16;
