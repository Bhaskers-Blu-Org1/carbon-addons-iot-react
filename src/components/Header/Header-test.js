import React from 'react';
import { mount } from 'enzyme';

import Header from './Header';

describe('Header testcases', () => {
  test('onClick', () => {
    const onClick = jest.fn();
    const wrapper = mount(
      <Header
        title="My Title"
        onClick={onClick}
        user="j@test.com"
        tenant="acme"
        appName="platform"
      />
    );
    wrapper.simulate('click');
    expect(onClick.mock.calls).toHaveLength(1);
  });
});
