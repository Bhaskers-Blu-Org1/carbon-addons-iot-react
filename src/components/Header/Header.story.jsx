import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import styled from 'styled-components';
import NotificationOn from '@carbon/icons-react/lib/notification/20';
import HeaderHelp from '@carbon/icons-react/lib/help/20';
import Avatar from '@carbon/icons-react/lib/user--avatar/20';

import Header from './Header';

React.Fragment = ({ children }) => children;

const StyledHeader = styled(Header)`
   {
    margin-top: 3rem;

    .bx--header__menu .bx--header__menu-item[role='menuitem'] {
      align-items: center;
      border: 0.125rem solid transparent;
      color: #f3f3f3;
      display: flex;
      padding: 1rem;
      height: 100%;
      font-size: 0.875rem;
      font-weight: 400;
      letter-spacing: 0;
      line-height: 1.125rem;
      text-decoration: none;
      user-select: none;
    }

    button.bx--header__menu-item {
      background: none;
      width: 100%;
    }

    .bx--text-truncate--end * {
      display: inline-block;
      vertical-align: middle;
    }
  }
`;

const User = styled.p`
   {
    color: white;
    font-size: 0.75rem;
    text-align: left;
    margin-right: 0.5rem;
    display: flex;
    flex-direction: column;
  }
`;

const HeaderProps = {
  user: 'JohnDoe@ibm.com',
  tenant: 'TenantId: Acme',
  className: 'custom-class-name',
  appName: 'Watson IoT Platform ',
  actionItems: [
    {
      label: 'alerts',
      onClick: action('pop alert modal'),
      btnContent: <NotificationOn fill="white" description="Icon" />,
    },
    {
      label: 'help',
      onClick: action('help drop open'),
      btnContent: (
        <HeaderHelp
          fill="white"
          description="Icon"
          className="bx--header__menu-item bx--header__menu-title"
        />
      ),
      childContent: [
        {
          metaData: {
            href: 'http://google.com',
            title: 'this is a title',
            target: '_blank',
            rel: 'noopener noreferrer',
            element: 'a',
          },
          content: 'this is my message to you',
        },
        {
          metaData: {
            onClick: action('do another action'),
            className: 'this',
            element: 'button',
          },
          content: (
            <span>
              JohnDoe@ibm.com
              <Avatar fill="white" description="Icon" />
            </span>
          ),
        },
      ],
    },
    {
      label: 'user',
      onClick: action('click'),
      btnContent: (
        <React.Fragment>
          <User>
            JohnDoe@ibm.com<span>TenantId: Acme</span>
          </User>
          <Avatar fill="white" description="Icon" />
        </React.Fragment>
      ),
      childContent: [
        {
          metaData: {
            href: 'http://google.com',
            title: 'this is a title',
            target: '_blank',
            rel: 'noopener noreferrer',
            element: 'a',
          },
          content: 'this is my message to you',
        },
        {
          metaData: {
            onClick: action('do another action'),
            className: 'this',
            element: 'button',
          },
          content: (
            <React.Fragment>
              JohnDoe@ibm.com
              <Avatar fill="white" description="Icon" />
            </React.Fragment>
          ),
        },
      ],
    },
  ],
};

storiesOf('Watson IoT|Header', module)
  .add('Header action buttons with dropdowns', () => <StyledHeader {...HeaderProps} />)
  .add('Header no submenu', () => (
    <StyledHeader
      {...HeaderProps}
      actionItems={[
        {
          label: 'user',
          onClick: action('click'),
          btnContent: (
            <React.Fragment>
              <User>
                JohnDoe@ibm.com<span>TenantId: Acme</span>
              </User>
              <Avatar fill="white" description="Icon" />
            </React.Fragment>
          ),
        },
      ]}
    />
  ));
