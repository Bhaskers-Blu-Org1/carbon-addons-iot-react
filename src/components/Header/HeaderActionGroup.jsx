import React, { useState } from 'react';
import { HeaderGlobalBar } from 'carbon-components-react/lib/components/UIShell';
import PropTypes from 'prop-types';

import HeaderAction from './HeaderAction/HeaderAction';
import { HeaderActionItemPropTypes } from './Header';

const propTypes = {
  actionItems: PropTypes.arrayOf(PropTypes.shape(HeaderActionItemPropTypes)).isRequired,
};

/**
 * Keeps track of the state of which header menu item is currently expanded
 *
 * Renders all the actions that can be clicked to navigate, open header panels (side panels),
 * or dropdown menus, passing an onToggleExpansion to each action
 */
const HeaderActionGroup = ({ actionItems }) => {
  const [expandedItem, setExpandedItem] = useState({});

  // expanded state for header dropdowns
  const toggleExpandedState = index => {
    setExpandedItem({
      [index]: !expandedItem[index],
    });
  };

  return (
    <>
      <HeaderGlobalBar>
        {actionItems.map((item, i) => (
          <HeaderAction
            item={item}
            index={i}
            onToggleExpansion={toggleExpandedState}
            isExpanded={expandedItem[item.label]}
            key={`header-action-item-${item.label}-${i}`}
          />
        ))}
      </HeaderGlobalBar>
    </>
  );
};

HeaderActionGroup.propTypes = propTypes;

export default HeaderActionGroup;
