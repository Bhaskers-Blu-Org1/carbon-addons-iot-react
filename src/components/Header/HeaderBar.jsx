import React, { useState } from 'react';
import { HeaderGlobalBar } from 'carbon-components-react/lib/components/UIShell';
import PropTypes from 'prop-types';

import HeaderAction from './HeaderAction/HeaderAction';
import { HeaderActionItemPropTypes, HeaderPanelPropTypes } from './Header';

const propTypes = {
  actionItems: PropTypes.arrayOf(HeaderActionItemPropTypes).isRequired,
  headerPanel: HeaderPanelPropTypes,
};

const defaultProps = {
  headerPanel: null,
};

/**
 * Toggles the state of which item is currently expanded
 * Does not listen for blur events, only expandedItem state changes
 * Contains actions that can be clicked to navigate, open header panels (side panels),
 * or dropdown menus
 */
const HeaderBar = ({ actionItems, headerPanel }) => {
  const [expandedItem, setExpandedItem] = useState({});

  // expanded state for header dropdowns
  const toggleExpandedState = index => {
    setExpandedItem({
      [index]: !expandedItem[index],
    });
  };

  return (
    <>
      <HeaderGlobalBar>
        {actionItems.map((item, i) => (
          <HeaderAction
            item={item}
            index={i}
            onToggleExpansion={toggleExpandedState}
            isExpanded={expandedItem[item.label]}
          />
        ))}
        {/* {headerPanel && (
          <>
            <HeaderAction />
            <HeaderGlobalAction
              aria-label="header-panel-trigger"
              key={appSwitcher}
              onClick={() => toggleExpandedState(appSwitcher)}
              title={appSwitcher}
              onBlur={e => toggleCloseOnTab(e)}
            >
              <AppSwitcher fill="white" description="Icon" />
            </HeaderGlobalAction>
            <HeaderPanel
              aria-label="Header Panel"
              className={cn(`${carbonPrefix}--app-switcher`, {
                [headerPanel.className]: headerPanel.className,
              })}
              expanded={expandedItem[appSwitcher]}
              data-testid="app-switcher-header-panel"
              onBlur={e => toggleClickOutside(e, appSwitcher)}
            >
              <headerPanel.content />
            </HeaderPanel>
          </>
        )} */}
      </HeaderGlobalBar>
    </>
  );
};

HeaderBar.propTypes = propTypes;
HeaderBar.defaultProps = defaultProps;

export default HeaderBar;
