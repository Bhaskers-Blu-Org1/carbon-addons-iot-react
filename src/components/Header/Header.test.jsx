import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Avatar from '@carbon/icons-react/lib/user--avatar/20';
import HeaderHelp from '@carbon/icons-react/lib/help/20';
import '@testing-library/jest-dom/extend-expect';

import { keyCodes } from '../../constants/KeyCodeConstants';

import Header, { APP_SWITCHER } from './Header';

React.Fragment = ({ children }) => children;

describe('Header testcases', () => {
  const onClick = jest.fn();
  const HeaderProps = {
    user: 'JohnDoe@ibm.com',
    tenant: 'TenantId: Acme',
    url: 'http://localhost:8989',
    className: 'custom-class-name',
    appName: 'Watson IoT Platform ',
    skipto: 'skip',
    actionItems: [
      {
        label: 'help',
        hasHeaderPanel: true,
        btnContent: (
          <HeaderHelp
            fill="white"
            description="Icon"
            className="bx--header__menu-item bx--header__menu-title"
          />
        ),
        childContent: [
          {
            onCLick: () => console.log('hi'),
            content: <p>This is a link</p>,
          },
          {
            onCLick: () => console.log('hi'),
            content: (
              <React.Fragment>
                <span>
                  JohnDoe@ibm.com
                  <Avatar fill="white" description="Icon" />
                </span>
              </React.Fragment>
            ),
          },
        ],
      },
      {
        label: 'user',
        btnContent: <Avatar fill="white" description="Icon" />,
        menuLinkName: 'a menu link name',
        childContent: [
          {
            metaData: {
              href: 'http://google.com',
              title: 'this is a title',
              target: '_blank',
              rel: 'noopener noreferrer',
              element: 'a',
            },
            content: 'this is my message to you',
          },
          {
            metaData: {
              onClick,
              className: 'this',
              element: 'button',
            },
            content: (
              <span>
                JohnDoe@ibm.com
                <Avatar fill="white" description="Icon" />
              </span>
            ),
          },
        ],
      },
    ],
  };

  const HeaderPropsWithoutOnClick = {
    user: 'JohnDoe@ibm.com',
    tenant: 'TenantId: Acme',
    url: 'http://localhost:8989',
    className: 'custom-class-name',
    appName: 'Watson IoT Platform ',
    skipto: 'skip',
    actionItems: [
      {
        label: 'user',
        onCLick: undefined,
        btnContent: <Avatar fill="white" description="Icon" />,
      },
    ],
  };

  it('should render', () => {
    const { container } = render(<Header {...HeaderProps} />);
    expect(container).toMatchSnapshot();
  });

  it('sidepanel should not render', () => {
    const { queryByLabelText } = render(<Header {...HeaderProps} />);
    expect(queryByLabelText(APP_SWITCHER)).toBeFalsy();
  });

  it('sidepanel should render', () => {
    const headerPanel = {
      className: 'header-panel',

      content: React.forwardRef((props, ref) => (
        // eslint-disable-next-line jsx-a11y/anchor-is-valid
        <a href="#" ref={ref} {...props}>
          Header panel content
        </a>
      )),
    };
    const { queryAllByLabelText } = render(<Header {...HeaderProps} headerPanel={headerPanel} />);
    expect(queryAllByLabelText(APP_SWITCHER).length).toBeGreaterThan(0);
  });

  it('children should render inside UL', () => {
    const { getByText } = render(<Header {...HeaderProps} />);
    expect(getByText('This is a link').parentElement.parentElement.nodeName).toBe('LI');
    expect(getByText('This is a link').parentElement.parentElement.className).toEqual(
      'action-btn__headerpanel-li'
    );
    expect(
      getByText('This is a link').parentElement.parentElement.parentNode.childElementCount
    ).toEqual(2);
  });

  it('clicking trigger button should expand the header panel', () => {
    const { getByTitle } = render(<Header {...HeaderProps} />);
    fireEvent.click(getByTitle('help'));
    expect(getByTitle('help').parentNode.lastChild.className).toContain(
      'bx--header-panel bx--header-panel--expanded action-btn__headerpanel'
    );
    fireEvent.click(getByTitle('help'));
    expect(getByTitle('help').parentNode.lastChild.className).toContain(
      'bx--header-panel action-btn__headerpanel action-btn__headerpanel--closed'
    );
  });

  it('closes when focus leaves panel', () => {
    const { getByTitle, getByText } = render(<Header {...HeaderProps} />);
    fireEvent.click(getByTitle('help'));
    fireEvent.focus(getByText('This is a link'));
    fireEvent.blur(getByText('This is a link'));
    expect(getByTitle('help').parentNode.lastChild.className).toContain(
      'bx--header-panel action-btn__headerpanel action-btn__headerpanel--closed'
    );
  });

  it('closes when focus leaves trigger', () => {
    const { getByTitle } = render(<Header {...HeaderProps} />);
    fireEvent.click(getByTitle('help'));
    expect(getByTitle('help').parentNode.lastChild.className).toContain(
      'bx--header-panel bx--header-panel--expanded action-btn__headerpanel'
    );
    fireEvent.blur(getByTitle('help'));
    expect(getByTitle('help').parentNode.lastChild.className).toContain(
      'bx--header-panel action-btn__headerpanel action-btn__headerpanel--closed'
    );
  });

  it('closes when focus leaves the current action', () => {
    const { getByTitle } = render(<Header {...HeaderProps} />);

    fireEvent.click(getByTitle('help'));
    expect(getByTitle('help').parentNode.lastChild.className).toContain(
      'bx--header-panel bx--header-panel--expanded action-btn__headerpanel'
    );
    // focus leaves the first button
    fireEvent.blur(getByTitle('help'));
    expect(getByTitle('help').parentNode.lastChild.className).toContain(
      'bx--header-panel action-btn__headerpanel action-btn__headerpanel--closed'
    );
  });

  it('App switcher opens', () => {
    const headerPanel = {
      className: 'header-panel',
      // eslint-disable-next-line react/no-multi-comp
      content: React.forwardRef((props, ref) => (
        // eslint-disable-next-line jsx-a11y/anchor-is-valid
        <a href="#" ref={ref} {...props}>
          Header panel content
        </a>
      )),
    };
    const { getByTitle } = render(<Header {...HeaderProps} headerPanel={headerPanel} />);

    fireEvent.click(getByTitle(APP_SWITCHER));
    expect(getByTitle(APP_SWITCHER).parentNode.lastChild.className).toContain(
      'bx--header-panel bx--header-panel--expanded bx--app-switcher'
    );

    fireEvent.click(getByTitle(APP_SWITCHER));
    expect(getByTitle(APP_SWITCHER).parentNode.lastChild.className).toContain(
      'bx--header-panel bx--app-switcher'
    );
  });

  it('onClick expands', () => {
    const headerPanel = {
      className: 'header-panel',
      // eslint-disable-next-line react/no-multi-comp
      content: React.forwardRef((props, ref) => (
        // eslint-disable-next-line jsx-a11y/anchor-is-valid
        <a href="#" ref={ref} {...props}>
          Header panel content
        </a>
      )),
    };
    const { getByTestId } = render(<Header {...HeaderProps} headerPanel={headerPanel} />);

    const menuItem = getByTestId('menuitem');
    fireEvent.click(menuItem);
    expect(menuItem.getAttribute('aria-expanded')).toBeTruthy();
  });

  it('onKeyDown expands with enter', () => {
    const headerPanel = {
      className: 'header-panel',
      // eslint-disable-next-line react/no-multi-comp
      content: React.forwardRef((props, ref) => (
        // eslint-disable-next-line jsx-a11y/anchor-is-valid
        <a href="#" ref={ref} {...props}>
          Header panel content
        </a>
      )),
    };
    const { getByTestId } = render(<Header {...HeaderProps} headerPanel={headerPanel} />);
    const menuTrigger = getByTestId('menuitem');
    fireEvent.keyDown(menuTrigger, { keyCode: keyCodes.ENTER });
    expect(menuTrigger.getAttribute('aria-expanded')).toBe('true');
    fireEvent.keyDown(menuTrigger, { keyCode: keyCodes.SPACE });
    expect(menuTrigger.getAttribute('aria-expanded')).toBe('false');
  });

  it('onKeyDown expands with space', () => {
    const headerPanel = {
      className: 'header-panel',
      // eslint-disable-next-line react/no-multi-comp
      content: React.forwardRef((props, ref) => (
        // eslint-disable-next-line jsx-a11y/anchor-is-valid
        <a href="#" ref={ref} {...props}>
          Header panel content
        </a>
      )),
    };
    const { getByTitle } = render(<Header {...HeaderProps} headerPanel={headerPanel} />);
    const menuTrigger = getByTitle('help');
    fireEvent.keyDown(menuTrigger, { keyCode: keyCodes.SPACE });
    expect(menuTrigger.getAttribute('aria-expanded')).toBe('true');
    fireEvent.keyDown(menuTrigger, { keyCode: keyCodes.SPACE });
    expect(menuTrigger.getAttribute('aria-expanded')).toBe('false');
  });

  it('onKeyDown esc on parent closes an open menu', () => {
    const headerPanel = {
      className: 'header-panel',
      // eslint-disable-next-line react/no-multi-comp
      content: React.forwardRef((props, ref) => (
        // eslint-disable-next-line jsx-a11y/anchor-is-valid
        <a href="#" ref={ref} {...props}>
          Header panel content
        </a>
      )),
    };
    const { getByTestId, getByRole } = render(
      <Header {...HeaderProps} headerPanel={headerPanel} />
    );
    const menuParent = getByRole('menu');
    const menuTrigger = getByTestId('menuitem');
    fireEvent.keyDown(menuTrigger, { keyCode: keyCodes.ENTER });
    expect(menuTrigger.getAttribute('aria-expanded')).toBe('true');
    fireEvent.keyDown(menuParent, { keyCode: keyCodes.ESCAPE });
    expect(menuTrigger.getAttribute('aria-expanded')).toBe('false');
  });

  it('onClick event on empty onClick prop', () => {
    const { getByLabelText } = render(<Header {...HeaderPropsWithoutOnClick} />);
    const menuItem = getByLabelText('user');
    fireEvent.click(menuItem);
    expect(getByLabelText('user')).toBeTruthy();
  });
});
