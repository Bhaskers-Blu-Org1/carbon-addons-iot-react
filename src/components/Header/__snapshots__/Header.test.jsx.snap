// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header testcases should render 1`] = `
<Header__StyledHeader
  aria-label="main header"
  className="main-header"
>
  <SkipToContent
    href="#main-content"
    tabIndex="0"
  >
    Skip to main content
  </SkipToContent>
  <HeaderMenuButton
    aria-label="Open menu"
    onClick={null}
  />
  <HeaderName
    href="#"
    prefix="IBM"
  >
    platform
  </HeaderName>
  <HeaderGlobalBar>
    <Header__StyledGlobalAction
      aria-label="alerts"
      key="menu-item-alerts-global"
      onClick={[MockFunction]}
    >
      <ForwardRef(Notification20)
        description="Icon"
        fill="white"
      />
    </Header__StyledGlobalAction>
    <ForwardRef
      aria-label="help"
      isMenu={false}
      key="menu-item-help"
      menuLinkName=""
      renderMenuContent={[Function]}
    >
      <ForwardRef(HeaderMenuItem)
        key="menu-item-help0-child"
      >
        <p>
          This is a link
        </p>
      </ForwardRef(HeaderMenuItem)>
      <ForwardRef(HeaderMenuItem)
        key="menu-item-help1-child"
      >
        <Component>
          <span>
            JohnDoe@ibm.com
            <ForwardRef(UserAvatar20)
              description="Icon"
              fill="white"
            />
          </span>
        </Component>
      </ForwardRef(HeaderMenuItem)>
    </ForwardRef>
  </HeaderGlobalBar>
</Header__StyledHeader>
`;
